keps/sig-instrumentation/2831-kubelet-tracing/README.md

```diff
@@ -0,0 +1,366 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+This Kubernetes Enhancement Proposal (KEP) is to enhance the kubelet to allow tracing gRPC and HTTP API requests.
+The kubelet is the integration point of a node's operating system and kubernetes. From [control plane-node communication documentation](https://kubernetes.io/docs/concepts/architecture/control-plane-node-communication/#control-plane-to-node),
+a primary communication path from the control plane to the nodes is from the apiserver to the kubelet process running on each node.
+The kubelet communicates with the container runtime over gRPC, where kubelet is the gRPC client and the Container Runtime Interface is the gRPC server.
+The CRI then sends the creation request to the container runtime installed on the node.
+Traces gathered from the kubelet will provide critical information to monitor and troubleshoot interactions at the node level.
+This KEP proposes using OpenTelemetry libraries, and exports in the OpenTelemetry format.
+This is in line with the [API Server enhancement](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/647-apiserver-tracing).
+
+## Motivation
+
+Along with metrics and logs, traces are a useful form of telemetry to aid with debugging incoming requests.
+The kubelet can make use of distributed tracing to improve the ease of use and enable easier analysis of trace data.
+Trace data is structured, providing the detail necessary to debug requests across service boundaries.
+As more core components are instrumented, Kubernetes becomes easier to monitor, manage, and troubleshoot.
+
+### Definitions
+
+**Span**: The smallest unit of a trace.  It has a start and end time. Spans are the building blocks of a trace.    
+**Trace**: A collection of Spans which represents work being done by a system. A record of the path of requests through a system.    
+**Trace Context**: A reference to a Trace that is designed to be propagated across component boundaries.    
+
+### Goals
+
+* The kubelet generates and exports spans for reconcile loops it initiates and for outgoing requests to the CRI. An example of a reconcile loop is the
+creation of a pod. Pod creation involves pulling the image, creating the pod sandbox and creating the container. Those would be linked together into a
```

Comment:
Will the Motivation above, the Goals here and Stories below also cover the various interactions between Kubelet and in-tree + CSI volume plugins (besides CRI)? If so, it will be great if that can be called out. For stateful workloads, attachment and mounting of volumes referred by a pod can be a significant contributor to latencies. Therefore being able to collect traces for a certain PV or ephemeral volume like a Secret as it gets mounted by different pods over the lifetime of the PV can be very useful!

Reply:
thanks, I've added a note about tracing wrt kubelet & volume plugins, ptal, and offer suggestions if it doesn't look right, thank you!

Reply:
@dashpole ptal, added information in `Goals` and `Example Scenarios` thanks

Reply:
In the first sentence, I think you can just enumerate more of the kubelet's interfaces, since we should try to instrument all of them.  E.g. 

"The kubelet generates and exports spans for reconcile loops it initiates and for incoming/outgoing requests to the kubelet's authenticated http servers, and the CRI, CNI, and CSI interfaces."

Reply:
updated, right.